core/metadata_test.cc

```diff
@@ -0,0 +1,198 @@
+    ModuleBuilder::all_module_builders_holder(true);
+  }
+
+  Module *m0;
+  Module *m1;
+};
+
+TEST(Metadata, RegisterSizeMismatchFails) {
+  struct bess::metadata::mt_attr attr0_s1 = {
+      .name = "attr0", .size = 1, .mode = bess::metadata::AccessMode::READ,
+  };
+  struct bess::metadata::mt_attr attr0_s2 = {
+      .name = "attr0", .size = 2, .mode = bess::metadata::AccessMode::WRITE,
+  };
+
+  ASSERT_EQ(0, default_pipeline.RegisterAttribute(&attr0_s1));
+  ASSERT_EQ(-EEXIST, default_pipeline.RegisterAttribute(&attr0_s2));
+}
+
+TEST_F(MetadataTest, DisconnectedFails) {
+  ASSERT_EQ(0,
+            m0->AddMetadataAttr("attr0", 1, bess::metadata::AccessMode::WRITE));
+  ASSERT_EQ(0,
+            m1->AddMetadataAttr("attr0", 1, bess::metadata::AccessMode::READ));
+  ASSERT_EQ(0, default_pipeline.ComputeMetadataOffsets());
+  ASSERT_TRUE(m1->attr_offsets[0] < 0);
+}
+
+TEST_F(MetadataTest, SingleAttrSimplePipe) {
+  ASSERT_EQ(0,
+            m0->AddMetadataAttr("attr0", 1, bess::metadata::AccessMode::WRITE));
+  ASSERT_EQ(0,
+            m1->AddMetadataAttr("attr0", 1, bess::metadata::AccessMode::READ));
+  m0->ConnectModules(0, m1, 0);
+
+  ASSERT_EQ(0, default_pipeline.ComputeMetadataOffsets());
+  ASSERT_EQ(0, m1->attr_offsets[0]);
+
+  for (size_t i = 0; i < m0->num_attrs; i++) {
+    ASSERT_EQ(m0->attr_offsets[i], m1->attr_offsets[i]);
+  }
+}
+
+TEST_F(MetadataTest, SingleAttrSimplePipeBackwardsFails) {
+  ASSERT_EQ(0,
+            m0->AddMetadataAttr("attr0", 1, bess::metadata::AccessMode::READ));
+  ASSERT_EQ(0,
+            m1->AddMetadataAttr("attr0", 1, bess::metadata::AccessMode::WRITE));
+  m0->ConnectModules(0, m1, 0);
+
+  ASSERT_EQ(0, default_pipeline.ComputeMetadataOffsets());
+  ASSERT_TRUE(m1->attr_offsets[0] < 0);
+}
+
+TEST_F(MetadataTest, MultipeAttrSimplePipe) {
+  bool dummy_meta[bess::metadata::kMetadataTotalSize] = {};
+  ASSERT_EQ(0, m0->AddMetadataAttr("a", 2, bess::metadata::AccessMode::WRITE));
+  ASSERT_EQ(1, m0->AddMetadataAttr("b", 3, bess::metadata::AccessMode::WRITE));
+  ASSERT_EQ(2, m0->AddMetadataAttr("c", 5, bess::metadata::AccessMode::WRITE));
+  ASSERT_EQ(3, m0->AddMetadataAttr("d", 8, bess::metadata::AccessMode::WRITE));
+  ASSERT_EQ(0, m1->AddMetadataAttr("a", 2, bess::metadata::AccessMode::READ));
+  ASSERT_EQ(1, m1->AddMetadataAttr("b", 3, bess::metadata::AccessMode::READ));
+  ASSERT_EQ(2, m1->AddMetadataAttr("c", 5, bess::metadata::AccessMode::READ));
+  ASSERT_EQ(3, m1->AddMetadataAttr("d", 8, bess::metadata::AccessMode::READ));
+  m0->ConnectModules(0, m1, 0);
+
+  ASSERT_EQ(0, default_pipeline.ComputeMetadataOffsets());
+
+  for (size_t i = 0; i < m0->num_attrs; i++) {
+    if (m0->attrs[i].mode == bess::metadata::AccessMode::READ)
+      continue;
+
+    bess::metadata::mt_offset_t offset = m0->attr_offsets[i];
+    for (mt_offset_t j = 0; j < m0->attrs[i].size; j++) {
+      ASSERT_FALSE(dummy_meta[offset + j]);
+      dummy_meta[offset + j] = true;
```

Comment:
Maybe I'm not totally getting the aim of this test -- I'm curious whether you want to be checking the actual offset results after having computed them above.